io/src/main/scala/sbt/io/Path.scala

```diff
@@ -602,3 +602,65 @@ private class ExcludeFiles(include: PathFinder, exclude: PathFinder) extends Pat
   private def _include: PathFinder = include
   private def _exclude: PathFinder = exclude
 }
+
+/**
+ * Represents a filtered subtree of the file system.
+ */
+sealed trait Glob {
+
+  /**
+   * The root of the file system subtree.
+   */
+  def base: File
+
+  /**
+   * The filter to apply to elements found in the file system subtree.
+   * @return the filter.
+   */
+  def filter: FileFilter
+
+  /**
+   * The maximum depth of elements to traverse. A depth of -1 implies that this glob applies only
+   * to the root specified by [[base]]. A depth of zero implies that only immediate children of
+   * the root are included in this glob. For positive depth, files may be included so long as their
+   * pathname, relativized with respect to the base, has no more than `depth + 1` components.
+   * @return the maximum depth.
+   */
+  def depth: Int
+}
+object Glob {
+  def apply(base: File, filter: FileFilter, depth: Int): Glob = new GlobImpl(base, filter, depth)
+  private class GlobImpl(val base: File, val filter: FileFilter, val depth: Int) extends Glob {
+    override def toString: String =
+      s"Glob(\n  base = $base,\n  filter = $filter,\n  depth = $depth\n)"
+    override def equals(o: Any): Boolean = o match {
+      case that: Glob =>
+        this.base == that.base && this.depth == that.depth && this.filter == that.filter
+      case _ => false
+    }
+    override def hashCode: Int = (((base.hashCode * 31) ^ filter.hashCode) * 31) ^ depth
+  }
+  final class GlobBuilder(val file: syntax.File) extends AnyVal {
```

Comment:
Should this be `File` as opposed to `syntax.File`?

Reply:
Yeah, though I think it doesn't really matter for most use cases since syntax.File is just an alias for File.